Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp

```diff
@@ -32,311 +32,259 @@
 {
+    if (m_didCompleteMessage) {
+        m_state = State::Terminating;
+        return true;
+    }
+
+    return false;
+}
+
+CurlMultipartHandle::FindBoundaryResult CurlMultipartHandle::findBoundary()
+{
+    FindBoundaryResult result;
+
     auto contentLength = m_buffer.size();
+    const auto contentStartPtr = m_buffer.data();
+    const auto contentEndPtr = contentStartPtr + contentLength;
+
+    auto boundaryLength = m_boundary.length();
+    auto boundaryStartPtr = m_boundary.data();
+
+    auto matchedStartPtr = static_cast<const uint8_t*>(memmem(contentStartPtr, contentLength, boundaryStartPtr, boundaryLength));
+    if (!matchedStartPtr) {
+        if (!m_didCompleteMessage) {
+            // Not enough data to check the boundary (Temporarily retain "Initial CRLF + (boundary - 1)" bytes for the next search.)
+            result.dataEnd = std::max(int(contentLength) - int(boundaryLength) + 1 - 2, 0);
+        } else
+            result.dataEnd = contentLength;
+
+        result.processed = result.dataEnd;
+        return result;
+    }
 
-    boundaryStartPosition = notFound;
-    lastPartialMatchPosition = contentLength;
+    auto matchedEndPtr = matchedStartPtr + boundaryLength;
 
-    if (contentLength < m_boundary.length())
-        return false;
+    // The initial CRLF is considered to be attached to the boundary delimiter line rather than
+    // part of the preceding part. (See RFC2046 [5.1.1. Common Syntax])
+    if (matchedStartPtr - contentStartPtr >= 2 && !memcmp(matchedStartPtr - 2, "\r\n", 2))
+        matchedStartPtr -= 2;
+    else if (matchedStartPtr - contentStartPtr >= 1 && !memcmp(matchedStartPtr - 1, "\n", 1))
+        matchedStartPtr--;
 
-    const auto content = m_buffer.data();
+    result.dataEnd = matchedStartPtr - contentStartPtr;
+    result.processed = result.dataEnd;
 
-    for (size_t i = 0; i < contentLength - m_boundary.length(); ++i) {
-        auto length = matchedLength(content + i);
-        if (length == m_boundary.length()) {
-            boundaryStartPosition = i;
-            return true;
+    // Check the Close Delimiter
+    if (contentEndPtr - matchedEndPtr < 2) {
+        if (!m_didCompleteMessage) {
+            // Not enough data to check the Close Delimiter.
+            return result;
         }
-
-        if (length)
-            lastPartialMatchPosition = i;
+    } else if (!memcmp(matchedEndPtr, "--", 2)) {
+        result.hasBoundary = true;
+        result.hasCloseDelimiter = true;
+        result.processed = matchedEndPtr + 2 - contentStartPtr;
+        return result;
     }
 
-    return false;
-}
-
-size_t CurlMultipartHandle::matchedLength(const uint8_t* data)
-{
-    auto length = m_boundary.length();
-    for (size_t i = 0; i < length; ++i) {
-        if (data[i] != m_boundary[i])
-            return i;
+    // Skip transport-padding
+    for (; matchedEndPtr < contentEndPtr && isTabOrSpace(*matchedEndPtr); ++matchedEndPtr) { }
```

Comment:
This code can't reliably skip white spaces if !m_didCompleteMessage. We need a new state to skip a blank link before headers.

Reply:
This process removes the transport-padding, not white spaces before headers.

https://datatracker.ietf.org/doc/html/rfc2046#section-5.1.1
```
   The boundary may be followed by zero or more characters of
   linear whitespace. It is then terminated by either another CRLF and
   the header fields for the next part, or by two CRLFs, in which case
   there are no header fields for the next part. 
```
```
     multipart-body := [preamble CRLF]
                       dash-boundary transport-padding CRLF
                       body-part *encapsulation
                       close-delimiter transport-padding
                       [CRLF epilogue]
```

Reply:
If the data is "... --boundary  " and not completed yet, and the next data is "   ", this code can't skip the following white spaces.

Reply:
In that case, I don't think there will be any problems because the following process will be executed and the next data will be waited for. Am I missing something?

https://github.com/WebKit/WebKit/pull/17403/files#diff-9748a6c95db83f4a038a66c392226087d69df3953308a3f96c081618d378e750R246-R248
```
    else if (matchedEndPtr >= contentEndPtr) {
        // Not enough data to check the boundary
        return result;
```

Reply:
The code doesn't keep `--boundary` in the buffer. And, it shouldn't keep. We need a new state.

Reply:
Why we shouldn't keep `--boundary` in the buffer? I think the current code is working fine. Why do we care so much about adding new states?

Reply:
That's because this code doesn't work in the case https://github.com/WebKit/WebKit/pull/17403#discussion_r1325424428
You shoud a test case of WebKitAPITests. layout test can't test all cases.

Reply:
I added a test case to CurlMultipartHandleTests.DivideTransportPadding.

Reply:
I got the idea why this code works. But, it's inefficient. Anyway, acceptable if you like this.

Reply:
I know that the current code is a little inefficient. But I want to keep the code simple, so I would like to use this code.